src/main/java/io/vertx/ext/stomp/impl/StompServerTCPConnectionImpl.java

```diff
@@ -101,7 +103,12 @@ public SSLSession sslSession() {
   @Override
   public void close() {
     cancelHeartbeat();
-    handler().onClose(this);
+    synchronized (this) {
+      if (!closed) {
+        handler().onClose(this);
+        closed = true;
+      }
+    }
```

Comment:
It would be better to avoid invoking callbacks from the synchronized block.

You could inspect the value of `closed` and store in a local boolean variable if action must be taken.

And then I think we should invoke `cancelHeartbeat` and `socket.close` only once too.

Reply:
Done.  and I added a test. How about this approach?